feat: 네트뭐크 모듈 추가

hryeong66 opened the pull request:
# What is this PR? 🔍
## 이슈
- #4 
## 설명
- 네트뭐크 모듈 추가했습니답
- 우선 example도 추가해두었습니다

# Changes 📝


# To Reviewers 🙏
EndPoint 관래서 고민이 되는 부분이 있는데여
Requestable에서 정의하는 프로퍼티들이 각 EndPoint에도 동일하게 들어갈거 같아서
```
public protocol Requestable {
  var url: String { get }
  var httpMethod: HTTPMethod { get }
  var headers: [String: String]? { get }
  var body: Data? { get }
  
  func buildURLRequest(with url: URL) -> URLRequest
}
```

1. EndPoint들이 Requestable을 채택하게 하는게 좋을까?
- NetworkRequest 객체를 따로 만들지 않고 EndPoint들이 Requestable을 바로 채택하는게 어떨까

2. Endpoint를 request 객체로 만드는게 좋을지
→ 이렇게 되면 https://github.com/mash-up-kr/MOIT-iOS/blob/develop/MOITNetwork/Interface/Endpoint.swift 
이런 방식으로 바꾸는게 좋지 않을까 합니당

의견 팍팍 얘기해주시면 감사드립니당🙇‍♀️

hryeong66 committed:
feat: #4 - 네트워크 모듈 추가

hryeong66 committed:
Etc: #4 - example 파일 추가

hryeong66 committed:
etc: 불필요한 주석 삭제

chansooo commented on Projects/Core/PPACModels/Sources/ExampleModel.swift:
```diff
@@ -0,0 +1,17 @@
+
+struct ExampleVO: Decodable {
+  let exampleString: String?
+}
```
도메인 로직에서 사용되는 엔티티의 경우는 추후에는 VO를 떼도 무방할 것 같습니다 ㅎㅎ

hryeong66 replied:
![image](https://github.com/mash-up-kr/PPAC-iOS/assets/59547116/7bd1aa67-9619-43b3-8339-6667f23818e8)
요 도식으로 보면 제가 처음에 생각했던게 responseModel → DTO, entity → VO로 생각했었는데
DTO, VO로 표현하는게 더 헷갈리는 것 같아서 ExampleResponseModel, ExampleEnity로 네이밍 변경했습니다 https://github.com/mash-up-kr/PPAC-iOS/pull/6/commits/a6a73b0f40f8b2acde4a9ab74fe926d67ee9d35c

jongnan replied:
이름을 풀어서 생각해보면 더 좋을꺼 같아요

DTO -> Data Transfer Object
VO -> Value Object

결국 둘 다 한 레이어에서 다른 레이어로 넘어갈 때 사용하는 것이고, 뮤터블하냐 이뮤터블하냐의 차이인거 같아요

entity가 중간에 변경이 될 수 있는 여지도 있다는 얘기를 하고 싶었어요!
그렇게 된다면, 엔티티 자체가 다른 레이어로 넘어갈텐데 DTO가 될 수도 있다는?

결국, 중요한건 Entity는 도메인 로직에서 사용하게 될 핵심 객체들인 것이고 postfix를 붙히냐 마냐도 상관없지만 네이밍이 길어질 수 있는 여지도 있기 때문에 제 생각에는 안붙혀도 될꺼 같습니다.

chansooo commented on Projects/Core/PPACNetwork/Sources/APIService.swift:
```diff
@@ -0,0 +1,34 @@
+  
+  private init() {}
+  
+  func request<T: Decodable>(_ request: Requestable) async throws -> T? {
+    guard let encodedUrl = request.url.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
+          let url = URL(string: encodedUrl) else {
+      throw APIServiceError.urlEncodingError
+    }
+    
+    let urlRequest = request.buildURLRequest(with: url)
+    let (data, response) = try await URLSession.shared.data(for: urlRequest)
+    
+    guard let httpResponse = response as? HTTPURLResponse,
+          (200..<500) ~= httpResponse.statusCode else {
+      throw APIServiceError.serverError
+    }
+    
+    let decoder = JSONDecoder()
+    let decodedData = try decoder.decode(T.self, from: data)
+    return decodedData
+  }
+}
```
현재 코드에서는 200번 이외에는 serverError로 내려오기 때문에 200, 300, 400, 500으로 나누거나, associated value로 status code를 전달해야 사용하는 부분에서 어떤 에러가 발생했는지 알 수 있을 것 같습니다. .!

GREENOVER replied:
디코딩 에러에 대해서도 구체적으로 오류를 던지면 반환 타입도 옵셔널할 필요없기에 조금 더 명확하게 오류 처리도 네트워크단에서 감지할 수 있어 보입니다 😃

GREENOVER replied:
+ 실제 에러 타입만 반환하는것보다 어떤 에러였는지 디버깅을 명확히 하기 위해 status code뿐 아니라 reponse, data 등도 info로 담아 처리해주면 더 보완될것 같네유!

hryeong66 replied:
NetworkError를 세분화 하고 return 을 (data: T?, error: Error?)를 해서 
호출하는 쪽에서도 error를 받아서 확인할 수 있도록 수정했습니다!
제안 감사합니다!! 🙇‍♀️ https://github.com/mash-up-kr/PPAC-iOS/pull/6/commits/b1e342ac0e80661d1fe14bdd341e0d95af11e738

GREENOVER replied:
오호 👍 그럼 Result 타입으로 반환하면 좀 더 편리하지 않을까요~?

chansooo commented on Projects/Core/PPACNetwork/Sources/Requestable.swift:
```diff
@@ -0,0 +1,36 @@
+  }
+}
+
+
```
NetworkRequest를 Requestable로 추상화를 하면서 얻는 이점이 어떤 것이 있을까요??

hryeong66 replied:
NetworkRequest 구조체를 따로 가지고 있을 필요가 없는 것 같아서
NetworkRequest를 삭제하고 Endpoint가 Requestable 채택하도록 변경했습니다! https://github.com/mash-up-kr/PPAC-iOS/pull/6/commits/b1e342ac0e80661d1fe14bdd341e0d95af11e738

chansooo replied:
좋습니다~ 
다만 현재 상황에서 body가 아닌 query로 넘길 때를 대비해서 추가를 하는 것이 좋아보이네요

hryeong66 replied:
오메나.. 요부분은 그 전에 참고 코드로 알려주신 부분 참고해서 수정했습니다!
[988a6ae](https://github.com/mash-up-kr/PPAC-iOS/pull/6/commits/988a6aeb9287d5f621e0d24a053e3de89ac10d9d)

chansooo commented on Projects/Features/Home/Sources/Service/ExampleServiceEndpoints.swift:
```diff
@@ -0,0 +1,62 @@
+  
+  var requestTimeOut: Int {
+    return 20
+  }
```
여러개의 endpoint가 생길 때 중복으로 구현되는 부분이 어느 부분이고, network모듈로 넘겨줄 때 인터페이스로 분리가 되어 있을까요??
Endpoint를 추상화해서 Netowork 모듈에서 두고 Requestable, NetworkRequest와의 역할을 한번에 하면 어떨까요?

hryeong66 replied:
넵! NetworkRequest 구조체를 한번 더 만드는 과정이 불필요한 것 같아서
NetworkRequest를 삭제하고 우선은 Endpoint가 Requestable을 채택하도록 수정했습니답!
음....제안주신 것 생각해보니 아예 Requestable → Endpoint로 네이밍을 변경하는 것도 괜찮을 것 같은데 어떤가욤??

chansooo replied:
이름 자체는 requestable과 endpoint 둘 다 괜찮아 보입니다. 
구현해주신대로 requestable을 protocol로 두고, 해당 프로토콜을 채택하는 객체들을 endpoiint로 정의해도 괜찮구요.
편하신대로 해주시면 될 것 같습니다.

chansooo commented on Projects/Features/Home/Sources/Service/ExampleRepository.swift:
```diff
@@ -0,0 +1,30 @@
+      return exampleVO
+    }
+  }
+}
```
repository과 apiclient의 각각의 역할이 무엇일까요?? 왜 나눠져야 할까요???

hryeong66 replied:
요거는.. 제가 잘못짠거 같아여🥲 repository과 apiclient가 같은 역할을 하고 있어서 apiclient을 삭제하고 repository에서 apiService request를 호출하도록 수정했습니다! https://github.com/mash-up-kr/PPAC-iOS/pull/6/commits/f340e5aa0c3668016338123a9b7b60eb3b1f0be3

chansooo commented on Projects/Features/Home/Sources/Service/ExempleAPIClient.swift:
```diff
@@ -0,0 +1,33 @@
+        return try? await self.apiService.request(request)
+    }
+    
+}
```
repository 에서 해당 apiclient를 composition하고 사용하는 이유가 있을까요??? 두 객체의 역할의 차이가 있을까요???

hryeong66 replied:
위 코멘트처럼 두 객체의 역할 차이가 없어서... apiclient을 삭제했습니다!! https://github.com/mash-up-kr/PPAC-iOS/commit/f340e5aa0c3668016338123a9b7b60eb3b1f0be3

chansooo commented on Projects/Features/Home/Sources/Service/ExampleServiceable.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+import PPACModels
+
+protocol ExampleServiceable {
+  func fetchExample() async throws -> ExampleDTO?
+  func postExample(id: Int) async throws -> ExampleDTO?
+}
```
NetworkServiceable과 ExampleServiceable은 각각의 역할에 차이가 있어 보이는데, 명확하게 구별할 수 있는 방법이 있을까요???

jongnan replied:
단순 궁금증인데 프로토콜의 postfix로 able을 사용하는건가요?

hryeong66 replied:
얘기해주신 것 처럼 역할은 다른데 네이밍이 비슷해서 
ExampleServiceable → ExampleRepository로 네이밍 변경했습니다! 

@jongnan 넴~! 스위프트 가이드에 따르면, ~할 수 있는 protocol을 정의하는 경우에 
swift에서 Decoable, Equatable 처럼 able을 많이 붙이는 것 같아요!
https://www.swift.org/documentation/api-design-guidelines/
`Protocols that describe a capability should be named using the suffixes able, ible, or ing (e.g. Equatable, ProgressReporting).`

그래서 실제 api Request를 구현할 쪽을 `NetworkServiceable`
request 객체를 생성하는 쪽을 `Requestable`로 지었습니다!
이렇게 보니까 ExampleServiceable 자체 네이밍이 넘 구렸네여.. 🥲

chansooo reviewed (approved):
너무 수고 많으셨습니다~

고민하시는 부분은 어떻게 구현하는지에 따라 둘 다 괜찮게 보이는데요,
각각의 프로토콜과 객체가 하는 역할이 무엇인지, 중복되는 역할은 없는지를 고민하시고, 원하시는 방식대로 하셔도 좋을 듯 합니다 ㅎㅎ

아래 링크는 네트워크 모듈 코드인데, 혹시나 참고가 될까 싶어서 남겨드립니다.

https://github.com/GongGanGam/GongGanGam-iOS/tree/develop/Modules/Network/Sources

jongnan commented on Projects/Core/PPACNetwork/Sources/Utils/Dictionary+Extension.swift:
```diff
@@ -0,0 +1,22 @@
+
+public extension Dictionary where Key == String {
+  
+  func toJsonString() -> String? {
```
이 반대도 있으면 좋지 않을까요?

String.toObject? String.toDTO?

chansooo replied:
@jongnan 
String 자체를 DTO로 만드는 경우는 비즈니스 로직에서 어떤 부분에서 필요할까요???

jongnan replied:
아 body 파싱할 때 사용하면 되겠다고 생각했었는데요!

생각해보니까 Decodable로 정의하고 파서가 따로 있었네요!

jongnan commented on Projects/Core/PPACNetwork/Sources/Requestable.swift:
```diff
@@ -0,0 +1,36 @@
+}
+
+extension Requestable {
+  public func buildURLRequest(with url: URL) -> URLRequest {
```
해당 함수는 기본형에 해당하는건가요?
프로토콜을 여기서 구현하는 이유는 무엇인가요??

hryeong66 replied:
해당 프로토콜을 채택하는 곳에서 공통으로 사용할 수 있게 extension에서 구현했습니다!
swift에서는 프로토콜이나 상위 타입에 사용하고 싶은 공통 함수가 필요하다면 extension에서 주로 구현하는 것 같아요ㅎㅎ
```
public extension Double {
    var toString: String {
        return String(self)
    }
// UILabel 타입과 이를 상속받는 곳에서도 모두 사용가능, 해당 함수를 재정의하싶다면
// override  func strikeThrough() {} 구현 
public extension UILabel {
 func strikeThrough() {
        let attrString: NSAttributedString?
        if let attributedText {
            attrString = attributedText.strikeThrough(self.textColor)
        } else {
            attrString = self.text?.strikeThrough(self.textColor)
        }
        self.attributedText = attrString
    }
}
```

만약 채택하는 쪽에서 해당 함수를 오버라이딩처럼 쓰고 싶다면
override 키워드 없이 public func buildURLRequest(with url: URL) -> URLRequest  이 함수를 구현해서 쓰면 됩니당

chansooo replied:
@jongnan 
추상 클래스라 생각하심 될 듯 합니다 ㅎㅎ

jongnan replied:
아하 답변 감사해요~

chansooo committed:
Merge branch 'develop' into feature/#4-network-hyerjang

hryeong66 committed:
feat: #4 - 코드리뷰 반영 - 불필요한 네트워크 모듈 파일 삭제

hryeong66 committed:
feat: #4- 코드리뷰 반영 - 네이밍 변경

hryeong66 committed:
feat: #4 - exampleModel 이름 변경

hryeong66 commented:
선생님덜 리뷰 감사합니다ㅏ🙇‍♀️

전체적으로 각 파일의 필요성과 네이밍을 다시 수정해서 정리했습니다!
![image](https://github.com/mash-up-kr/PPAC-iOS/assets/59547116/ac972fd4-debf-4ff5-8541-aa1341e6aa17)

- 불필요하게 apiClient가 중간에 들어가 있어서 해당 클래스는 삭제하고 Respository에서 api request를 하도록 수정했습니답!
- 데이터 모델 네이밍은 responseModel → DTO, entity → VO로 생각했었는데
DTO, VO로 표현하는게 더 헷갈리는 것 같아서 ExampleResponseModel, ExampleEnity로 네이밍 변경했습니다~!

chansooo commented on Projects/Core/PPACNetwork/Sources/NetworkService.swift:
```diff
@@ -0,0 +1,44 @@
+
+import Foundation
+
+final class NetworkService: NetworkServiceable {
```
혹시 NetworkServicable로 추상화한 프로토콜을 두신 이유가 있을까요???

hryeong66 replied:
RepositoryImpl가 networkService를 가지고 있어야 하는데 
그때 타입을 프로토콜로 해서 주입할 수 있게 하는게 좋지 않을까 했습니다!

chansooo commented on Projects/Core/PPACNetwork/Sources/NetworkService.swift:
```diff
@@ -0,0 +1,44 @@
+
+final class NetworkService: NetworkServiceable {
+  
+  func request<T: Decodable>(_ request: Requestable) async -> (data: T?, error: Error?) {
+    guard let encodedUrl = request.url.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
+          let url = URL(string: encodedUrl) else {
+      return (nil, NetworkError.urlEncodingError)
+    }
+    
+    let urlRequest = request.buildURLRequest(with: url)
+    
+    let (data, response): (Data, URLResponse)
+    do {
+      (data, response) = try await URLSession.shared.data(for: urlRequest)
+    } catch {
+      return (nil, NetworkError.invalidResponse)
+    }
+    
+    guard let httpResponse = response as? HTTPURLResponse else {
+      return (nil, NetworkError.invalidResponse)
+    }
+    
+    switch httpResponse.statusCode {
+    case 200..<300:
+      let decoder = JSONDecoder()
+      let decodedData = try? decoder.decode(T.self, from: data)
+      return (decodedData, nil)
+    case 400..<500:
+      return (nil, NetworkError.clientError(statusCode: httpResponse.statusCode, message: String(data: data, encoding: .utf8)))
+    case 500..<600:
+      return (nil, NetworkError.serverError(statusCode: httpResponse.statusCode, message: String(data: data, encoding: .utf8)))
+    default:
+      return (nil, NetworkError.unknown)
+    }
+  }
```
반환 타입이 tuple로 되어 있을 경우 (data, error)의 경우의 수가
1. nil, not nil
2. nil nil
3. not nil, not nil
4. not nil, nil
총 4가지가 되게 됩니다.
이 중에서 2와 3은 일어날 수 없는 경우의 수이구요.
따라서 해당 부분을 Result 타입으로 변경해서 경우의 수를 2개로 줄이면 해당 객체를 사용하는 부분에서 핸들링을 하기 더 쉬워질 것 같습니다~

hryeong66 replied:
넵 해당 부분 Result 타입으로 변경했습니다 https://github.com/mash-up-kr/PPAC-iOS/pull/6/commits/4b3c5cf96e9f5f7dde96fef6b7f5e109bb47fe8e

chansooo commented on Projects/Core/PPACModels/Sources/ExampleModel.swift:
```diff
@@ -7,11 +7,11 @@
-struct ExampleVO: Decodable {
+struct ExampleEntity: Decodable {
   let exampleString: String?
 }
```
네이밍은 마음대로 해도 되지만 domain layer에서 사용되는 모델들의 경우는 entity를 떼도 좋을 듯 싶습니다.
이건 추후에 같이 컨벤션 정하면서 이야기해보시죠 ㅎㅎ
e.g. PhotoEntity -> Photo

jongnan replied:
https://github.com/mash-up-kr/PPAC-iOS/pull/6#discussion_r1640005411

위에서도 적었두었는데, 저도 찬수님의 의견의 동의합니다.

hryeong66 replied:
넵 좋습니다! 예시지만 entity 네이밍 제거했습니다 [f1236a8](https://github.com/mash-up-kr/PPAC-iOS/pull/6/commits/f1236a84b24696ba8b1cb78ffc179e065152ff93)

chansooo commented on Projects/Features/Home/Sources/Service/ExampleRepository.swift:
```diff
@@ -21,17 +21,17 @@ class ExampleRepositoryImpl: ExampleRepository {
     guard let data, error == nil else { return nil }
-    return ExampleVO(exampleString: response.exampleString ?? "")
+    return ExampleEntity(exampleString: data.exampleString ?? "")
   }
```
예시이긴 하지만, network 모듈 내부에서 에러를 분기처리해서 던졌는데, data layer에서 nil로 반환을 해버리면 의미가 퇴색되는 것 같습니다!

hryeong66 committed:
feat: #4 - Network logger 추가

hryeong66 committed:
feat: #4 - 코드리뷰 반영 - request return type Result로 변경

jongnan commented on Projects/Core/PPACNetwork/Sources/Utils/NetworkLogger.swift:
```diff
@@ -0,0 +1,51 @@
+    static func logError(_ error: NetworkError) {
+        switch error {
+        case .urlEncodingError:
+            print("❌ [ERROR]: URL Encoding Error")
+        case .dataDecodingError:
+            print("❌ [ERROR]: Data Decoding Error")
+        case .invalidResponse:
+            print("❌ [ERROR]: Invalid Response")
+        case .clientError(let statusCode, let message):
+            print("❌ [CLIENT ERROR]: StatusCode: \(statusCode), Message: \(message ?? "N/A")")
+        case .serverError(let statusCode, let message):
+            print("❌ [SERVER ERROR]: StatusCode: \(statusCode), Message: \(message ?? "N/A")")
+        case .unknown:
+            print("❌ [ERROR]: Unknown Error")
```
간단하게 서칭을 해봤는데요!

https://velog.io/@segassdc1/Logging%EC%97%90-%EB%8C%80%ED%95%9C-%EA%B3%A0%EC%B0%B0

이렇게 로거를 이용해서 에러를 찍는게 좀 더 디버깅할 때나 가시성에도 좋지 않을까요?

hryeong66 replied:
제안 감사합니다!! 요거 활용해서 수정해볼게요!

jongnan commented on Projects/Features/Home/Sources/Service/ExampleRepository.swift:
```diff
@@ -0,0 +1,41 @@
+import PPACModels
+import PPACNetwork
+
+protocol ExampleRepository {
```
여기에는 able을 안쓴 이유가 있나요?
Repositable?

chansooo replied:
able을 붙이는 이유가 하나의 동작에 대해서 protocol을 정의하고, 해당 동작을 수행하는 객체에 채택하는 용도인데요,
e.g. drawable이라는 protocol은 shape라는 필드와 draw()라는 함수를 가지도록.
즉, 동사의 뒤에 접미어가 붙는 것이 자연스럽습니다.
repository처럼 명사형인 경우에는 의미가 더 어색해질 수 있기 때문에 지금처럼 두고, 구현부를 ExampleRepositoryImpl 등으로 네이밍해도 괜찮을 것 같습니다.

hryeong66 committed:
feat: #4 - 코드리뷰 반영 request parameter 변경

feat: #4 - 코드리뷰 반영 request parameter 변경

feat: #4 - 코드리뷰 반영 request parameter 변경

hryeong66 committed:
feat: #4 - 코드리뷰 반영 - 모델명 entity 삭제

feat: #4 - 코드리뷰 반영 - 모델명 entity 삭제

hryeong66 commented on Projects/Core/PPACModels/Sources/ExampleModel.swift:
```diff
@@ -0,0 +1,29 @@
+    let container = try decoder.container(keyedBy: CodingKeys.self)
+    exampleString = try container.decodeIfPresent(String.self, forKey: .exampleString)
+  }
+}
```
요건 예시에서 Example(exampleString: String) init 호출하려구 추가했습니답
